api/src/sparql.rs

```diff
@@ -0,0 +1,63 @@
+//! Common traits for working with [SPARQL](https://www.w3.org/TR/sparql11-query/).
+//!
+//! # Design rationale
+//!
+//! These traits are deliberately very generic.
+//! Specific implementations may have additional features, such as:
+//! 
+//! - preparing a query for multiple use;
+//! - setting default values for `BASE`, `PREFIX`, `FROM`, `FROM NAMED` directives,
+//!   before parsing query string;
+//! - pre-binding variables before evaluating query;
+//! - etc...
+//!
+//! However, we do not want to impose these feature, or any subset thereof,
+//! to all implementation of Sophia.
+
+use crate::dataset::Dataset;
+use crate::graph::Graph;
+use crate::prefix::PrefixMap;
+use crate::term::TTerm;
+use std::error::Error;
+
+/// A dataset that can be queried with SPARQL.
+pub trait SparqlDataset: Dataset {
+    type Result: SparqlResult;
+    type SparqlError: Error + 'static;
+
+    /// Parse and immediately execute `query`
+    fn execute<'p, P>(&self, query: &str) -> Result<Self::Result, Self::SparqlError>
+    where
+        P: PrefixMap<'p> + ?Sized,
+    ;
+}
+
+/// The result of executing a SPARQL query.
+pub trait SparqlResult {
+    type Term: TTerm;
+    /// The type of iterator returned if the query was a SELECT.
+    type Bindings: Iterator<Item=Vec<Option<Self::Term>>>;
```

Comment:
This iterator does not keep track of the  variable name -> tuple position mapping that is useful to e.g. make sense of a "SELECT *" query results or serialize the query results.

Reply:
The order of the variables is accessible through `SparqlResult::variables`, and the binding iterators through `SparqlResult::bindings`. Attaching the variables to the iterator *itself* would require yet another trait, making the API more complex, without a clear (to me at last) added value.

Reply:
Indeed. I should have read the code more carefully. Sorry.

What about returning something like `Vec<&str>` or `&[&str]`? This way the implementation would not have to clone the data.

Reply:
> What about returning something like Vec<&str> or &[&str]? This way the implementation would not have to clone the data.

On the other hand, that would force *all* implementations to permanently own a copy of the variable names... Granted, that would not be such a big deal.

However, I chose to make `SparqlResult::bindings` "consume" (i.e. take ownership) of the `QueryResult`. That way, an implementation could be built to just "stream" the JSON returned by a SPARQL endpoint, and not have to store anything. So you need to get ownership of the variable names before calling `bindings()`. Hence the choice to make `variables()` return `String`s rather than `&str`s.

But that's a trade-off I am open to discuss if you think there are better options.

Reply:
Thank you for make `bindings` consume the `QueryResult`. It also seems to me that it is the best option.

After your reply, I agree with you about the `variables` method return type. Returning `Vec<String>` seems indeed the best way to be easy to implement. Thank you!